src/network/networkd-address.c

```diff
@@ -20,6 +21,85 @@
 #define ADDRESSES_PER_LINK_MAX 2048U
 #define STATIC_ADDRESSES_PER_NETWORK_MAX 1024U
 
+/* Unique Local IPv6 Unicast Addresses https://tools.ietf.org/rfc/rfc4193.txt */
+int generate_ipv6_ula_address(Link *link, uint16_t subnet_id, Address **ret) {
+        _cleanup_(address_freep) Address *address = NULL;
+        _cleanup_(khash_unrefp) khash *h = NULL;
+        uint8_t key[16], eui64[8], global_id[5];
+        const void *p;
+        uint64_t ntp;
+        int r;
+
+        r = address_new(&address);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not allocate address: %m");
+
+        address->family = AF_INET6;
+
+        /* 1. Obtain the current time of day in 64-bit NTP format [NTP]. */
+        ntp = (int64_t) now(CLOCK_MONOTONIC);
```

Comment:
It seems to me like depending on the monotonic clock is going to result in a different unique local address for each time the interface is brought up.

At least in the application that I was hoping to use ULAs for, that's actually not a great outcome. The goal was to get stable ULAs that don't change after being generated.

Reply:
Maybe the machine-id could be used for this?

Reply:
That's not in RFC.

Reply:
The point of the RFC is that the unique local address gets generated *once* by the administrator of the network.

The ULA is then assigned to a machine, and stays the same for the lifetime of the machine, or ULA.

Generating a ULA that changes every time the interface comes up defeats the point.

Reply:
Additionally, the point of the "global id" portion of the ULA is that it's chosen by an administrator, and then assigned to all machines controlled by that administrators organization.

Systemd shouldn't generate this, because it needs to be assigned by the admin.

As far as I can tell, the only portion of the ULA that should be determined by systemd is the interfaceid, which should be stable for a particular machine as well.

Reply:
this is not according to 3.2.2.  of RFC. May be some other feature  or may be another RFE.

Reply:
You need to re-read the RFC more carefully.
3.2.2 is not intended to be done on every machine. The point of 3.2.2 is *an example* of how a local administrator may generate a global Id with which to then generate staticly assigned ulas to machines in a network.

Generating a new global id every time the interface is brought up completely defeats the purpose of having ULAs

Reply:
Yeah, @jonesmz is correct in this case. Maybe, moving the logic to `networkctl` to add, e.g., `networkctl generate-ula-globalid`.

Reply:
hmm then no need to the calculating this. We just can assign. Why didn't you figure out his before ?

Reply:
Also see my comment below:
> BTW, because of this, we cannot share the global ID with other machines...

Reply:
> Why didn't you figure out this before ?

Sorry for late notice :-p

Reply:
> We just can assign.

Right.

Reply:
I think we need to supply the same in other machines too in the network. Hence it's kind user will supply